Source/WebCore/Modules/fetch/FetchHeaders.cpp

```diff
@@ -228,16 +268,46 @@ void FetchHeaders::filterAndFill(const HTTPHeaderMap& headers, Guard guard)
 std::optional<KeyValuePair<String, String>> FetchHeaders::Iterator::next()
 {
     if (m_keys.isEmpty() || m_updateCounter != m_headers->m_updateCounter) {
+        bool hasSetCookie = !m_headers->m_setCookieValues.isEmpty();
         m_keys.resize(0);
-        m_keys.reserveCapacity(m_headers->m_headers.size());
+        m_keys.reserveCapacity(m_headers->m_headers.size() + (hasSetCookie ? 1 : 0));
         for (auto& header : m_headers->m_headers)
             m_keys.uncheckedAppend(header.key.convertToASCIILowercase());
+        if (hasSetCookie)
+            m_keys.uncheckedAppend("set-cookie"_str);
         std::sort(m_keys.begin(), m_keys.end(), WTF::codePointCompareLessThan);
+
+        // We adjust the current index to work with Set-Cookie headers.
+        // This relies on the fact that `m_currentIndex + m_setCookieIndex`
+        // gives you the current total index into the iteration.
+        m_currentIndex += m_setCookieIndex;
+        if (hasSetCookie) {
+            size_t setCookieKeyIndex = std::lower_bound(m_keys.begin(), m_keys.end(), "set-cookie"_str, WTF::codePointCompareLessThan) - m_keys.begin();
+            if (m_currentIndex < setCookieKeyIndex)
+                m_setCookieIndex = 0;
+            else {
+                m_setCookieIndex = std::min(m_currentIndex - setCookieKeyIndex, m_headers->m_setCookieValues.size());
+                m_currentIndex -= m_setCookieIndex;
+            }
+        } else
+            m_setCookieIndex = 0;
+
         m_updateCounter = m_headers->m_updateCounter;
     }
     while (m_currentIndex < m_keys.size()) {
-        auto key = m_keys[m_currentIndex++];
-        auto value = m_headers->m_headers.get(key);
+        auto key = m_keys[m_currentIndex];
+        String value;
+        if (key == "set-cookie"_str) {
```

Comment:
It is a bit sad we are doing this check, this is not really efficient.
Can we instead have m_keys storing SetCookieOrOtherName? Or maybe consider the null string is Set-Cookie?

Reply:
If we treat the null string, or some other value, as `Set-Cookie`, we would have to use a custom comparison function when sorting and finding the `Set-Cookie` index. But I was wondering if there is some way to optimize this check with `AtomString` instead.

Reply:
The custom function would basically call codePointCompareLessThan except if a or b is null in which case it would call codePointCompareLessThan with a or b being set to "set-cookie".